Examples/DpdkBridge/main.cpp

```diff
@@ -0,0 +1,336 @@
+	createCoreVectorFromCoreMask(coreMaskToUse, coresToUse);
+
+	// need minimum of 2 cores to start - 1 management core + 1 (or more) worker thread(s)
+	if (coresToUse.size() < 2)
+	{
+		EXIT_WITH_ERROR("Needed minimum of 2 cores to start the application");
+	}
+
+	// initialize DPDK
+	if (!DpdkDeviceList::initDpdk(coreMaskToUse, mBufPoolSize))
+	{
+		EXIT_WITH_ERROR("Couldn't initialize DPDK");
+	}
+
+	// removing DPDK master core from core mask because DPDK worker threads cannot run on master core
+	coreMaskToUse = coreMaskToUse & ~(DpdkDeviceList::getInstance().getDpdkMasterCore().Mask);
+
+	// re-calculate cores to use after removing master core
+	coresToUse.clear();
+	createCoreVectorFromCoreMask(coreMaskToUse, coresToUse);
+
+	// collect the list of DPDK devices
+	vector<DpdkDevice*> dpdkDevicesToUse;
+	for (vector<int>::iterator iter = dpdkPortVec.begin(); iter != dpdkPortVec.end(); iter++)
+	{
+		DpdkDevice* dev = DpdkDeviceList::getInstance().getDeviceByPort(*iter);
+		if (dev == NULL)
+		{
+			EXIT_WITH_ERROR("DPDK device for port %d doesn't exist", *iter);
+		}
+		dpdkDevicesToUse.push_back(dev);
+	}
+
+	// go over all devices and open them
+	for (vector<DpdkDevice*>::iterator iter = dpdkDevicesToUse.begin(); iter != dpdkDevicesToUse.end(); iter++)
+	{
+		if (!(*iter)->openMultiQueues(1,1))
+		{
+			EXIT_WITH_ERROR("Couldn't open DPDK device #%d, PMD '%s'", (*iter)->getDeviceId(), (*iter)->getPMDName().c_str());
+		}
+	}
+
+	// get DPDK device to send packets to (or NULL if doesn't exist)
+	DpdkDevice* sendPacketsTo = DpdkDeviceList::getInstance().getDeviceByPort(sendPacketsToPort);
+	if (sendPacketsTo != NULL && !sendPacketsTo->isOpened() &&  !sendPacketsTo->open())
+	{
+		EXIT_WITH_ERROR("Could not open port#%d for sending matched packets", sendPacketsToPort);
+	}
+
+	// prepare configuration for every core
+	AppWorkerConfig workerConfigArr[2];
+	workerConfigArr[0].CoreId = coresToUse.at(0).Id;
+	workerConfigArr[0].RxDevice = dpdkDevicesToUse.at(0);
+	workerConfigArr[0].TxDevice = dpdkDevicesToUse.at(1);
+	workerConfigArr[1].CoreId = coresToUse.at(1).Id;
+	workerConfigArr[1].RxDevice = dpdkDevicesToUse.at(1);
+	workerConfigArr[1].TxDevice = dpdkDevicesToUse.at(0);
+
+	// create worker thread for every core
+	vector<DpdkWorkerThread*> workerThreadVec;
+	workerThreadVec.push_back(new AppWorkerThread(workerConfigArr[0]));
+	workerThreadVec.push_back(new AppWorkerThread(workerConfigArr[1]));
+
+	// start all worker threads
+	if (!DpdkDeviceList::getInstance().startDpdkWorkerThreads(coreMaskToUse, workerThreadVec))
+	{
+		EXIT_WITH_ERROR("Couldn't start worker threads");
+	}
+
+	// register the on app close event to print summary stats on app termination
+	DpdkBridgeArgs args;
+	args.workerThreadsVector = &workerThreadVec;
+	ApplicationEventHandler::getInstance().onApplicationInterrupted(onApplicationInterrupted, &args);
+
+	// infinite loop (until program is terminated)
+	while (!args.shouldStop)
+	{
+		sleep(5);
```

Comment:
You may put `sleep(1)` instead of  `sleep(5)` for better reaction

Reply:
I also would print here current traffic stats, consider a different alternative for the loop:

        uint64_t counter = 0;
        int statsCounter = 1;

        // Keep running while flag is on
        while (!args.shouldStop)
        {
                // Sleep for 1 second
                sleep(1);

                // Print stats every COLLECT_STATS_EVERY_SEC seconds
                if (counter % COLLECT_STATS_EVERY_SEC == 0)
                {
                        printf("\n\nStats #%d\n", statsCounter++);
                        printf("==========\n\n");

                        // Print stats of traffic going from Device1 to Device2
                        printf("\nDevice1->Device2 stats:\n\n");
                        printStats(device1, device2);

                        // Print stats of traffic going from Device2 to Device1
                        printf("\nDevice2->Device1 stats:\n\n");
                        printStats(device2, device1);
                }
                counter++;
        }

Here is the implementation of `printStats()`:

```
void printStats(pcpp::DpdkDevice* rxDevice, pcpp::DpdkDevice* txDevice)
{
        pcpp::DpdkDevice::DpdkDeviceStats rxStats;
        pcpp::DpdkDevice::DpdkDeviceStats txStats;
        rxDevice->getStatistics(rxStats);
        txDevice->getStatistics(txStats);

        std::vector<std::string> columnNames;
        columnNames.push_back(" ");
        columnNames.push_back("Total Packets");
        columnNames.push_back("Packets/sec");
        columnNames.push_back("Data");
        columnNames.push_back("Gbps");

        std::vector<int> columnLengths;
        columnLengths.push_back(10);
        columnLengths.push_back(15);
        columnLengths.push_back(15);
        columnLengths.push_back(15);
        columnLengths.push_back(15);

        pcpp::TablePrinter printer(columnNames, columnLengths);

        std::stringstream totalRx;
        totalRx << "rx" << "|" << rxStats.aggregatedRxStats.packets << "|" << rxStats.aggregatedRxStats.packetsPerSec << "|" << rxStats.aggregatedRxStats.bytes << "|" << rxStats.aggregatedRxSta$
        printer.printRow(totalRx.str(), '|');

        std::stringstream totalTx;
        totalTx << "tx" << "|" << txStats.aggregatedTxStats.packets << "|" << txStats.aggregatedTxStats.packetsPerSec << "|" << txStats.aggregatedTxStats.bytes << "|" << txStats.aggregatedTxSta$
        printer.printRow(totalTx.str(), '|');
}

```

Reply:
Agreed, I added, with some small modifications, to make it more like DPDK's l2fwd example, like grouping stats per device and not per flow (I think actually it more generic, in case we add support for a third device in the future), or cleaning the screen between iterations (is much more easier to follow the execution that way).